keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,857 @@
+webhook:
+
+```yaml
+rules:
+  - operations: ['CREATE']
+    apiGroups: '' # core
+    apiVersions: '*'
+    resources: 'pods'
+predicates:
+  # Only include pods with an NFS volume.
+  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
+```
+
+### Goals
+
+1. Provide a filtering mechanism for excluding requests from an admission webhook
+2. Maintain consistency with `ValidatingAdmissionPolicy`
+
+### Non-Goals
+
+* Provide a mechanism to exclude requests from all webhooks.
+
+## Proposal
+
+### API
+
+Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
+with a new `Predicates` field:
+
+```go
+
+type ValidatingWebhook struct {
+  // ...
+
+  // Predicates is a list of conditions on the AdmissionRequest ('request') that must be met for a
+  // request to be sent to this webhook.
+  // +optional
+  Predicates []Predicate `json:"predicates,omitempty"`
+}
+
+type MutatingWebhook struct {
+  // ...
+  Predicates []Predicate `json:"predicates,omitempty"`
+}
+
+// Predicate represents a condition which must by fulfilled for a request to be sent to a webhook.
+type Predicate struct {
+  // Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the AdmissionRequest, organized into CEL variables:
+	//
+	//'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// request.object. No other metadata properties are accessible.
+	//
+	// Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible.
+	// Accessible property names are escaped according to the following rules when accessed in the expression:
+	// - '__' escapes to '__underscores__'
+	// - '.' escapes to '__dot__'
+	// - '-' escapes to '__dash__'
+	// - '/' escapes to '__slash__'
+	// - Property names that exactly match a CEL RESERVED keyword escape to '__{keyword}__'. The keywords are:
+	//	  "true", "false", "null", "in", "as", "break", "const", "continue", "else", "for", "function", "if",
+	//	  "import", "let", "loop", "package", "namespace", "return".
+	// Examples:
+	//   - Expression accessing a property named "namespace": {"Expression": "object.__namespace__ > 0"}
+	//   - Expression accessing a property named "x-p
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+```
+
+The predicate expression has access to the contents of the `AdmissionRequest` object (exposed as the
```

Comment:
For consistency with ValidatingAdmissionPolicies, make 'object' a top level variable? This will also make it easier to consolidate code between policies and webhooks.

Reply:
+1 -- I would also make a case for `oldObject` to be a top level variable for mutation cases.